### DISCUSSION TITLE
Пустая веб страница на iOS (хостинг GitHub)

### DISCUSSION THREAD

**Discussion by efremych** [Chimildic/goofy#101] [Помощь с ошибкой] [answered]
Решил освежить по инструкции у себя страничку [управления с телефона](https://github.com/Chimildic/goofy/discussions/9#discussion-77428).
Обновил код, настроил хостинг на GitHub.

Вот только по ссылке GitHub страничка не грузится на ios. Пустая страничка.
С ПК и на андроиде - ок.

Может, какие-то дополнительные хитрости нужны на ios? Браузер - Хром.

**Comment by Chimildic** [comment]
> Идентификация по входу в гугл аккаунт из браузера, в котором открывается страница.

Попробуйте войти в аккаунт браузера или просто почту от гугла.

**Comment by efremych** [comment]
> Попробуйте войти в аккаунт браузера или просто почту от гугла.

Так я и не выходил оттуда уже много лет.

Вот два варианта - с гитхабом и без. В принципе не против и потерпеть блок сверху, но если это поправимо, было бы здорово

![IMG_2117](https://user-images.githubusercontent.com/76212904/124398076-69df5980-dd1c-11eb-9d08-a7f7cad41d2b.jpeg)
![IMG_2116](https://user-images.githubusercontent.com/76212904/124398077-6a77f000-dd1c-11eb-8e9a-7432e68515f6.PNG)

**Reply by Chimildic** [reply]
Откройте станицу `chrome://inspect`, снова зайдите на гитхаб версию. На первой странице будут логи?

**Reply by efremych** [reply]
Нет. Логи появляются если на работающую страницу заходить

**Reply by Chimildic** [reply]
У меня нет ios, чтобы воспроизвести. Если все по шагам инструкции правильно, то и не знаю куда копать. Другой браузер ради теста?

**Reply by efremych** [reply]
В Сафари еще смотрел - то же самое. Попробую, может, попозже в другом каком браузере, но, что-то, мне кажется, дело где-то в ios.
Рядом андроид вот лежит - всё четко, и на ПК тоже 2 вкладки открыто: одна с версией гитхаба, другая - старая верcия - тоже всё четко.

Но, конечно, крутая штука. Зря долго откладывал своё обновление, думал долго разбираться буду :) А оказалось и всё понятно по инструкции, и в принципе все нужные функции быстренько запулил в телефон. И понятно, что и, самое главное, как еще запилить туда можно будет. Спасибо :)

![image](https://user-images.githubusercontent.com/76212904/124399777-6e107480-dd26-11eb-894f-eb5bb6d97870.png)

**Reply by efremych** [reply]
Попробовал еще в Яндекс-браузере и в Opera - тот же результат. Надо, конечно, чтоб еще кто-нибудь проверил для чистоты эксперимента...

**Reply by efremych** [reply]
Извиняюсь за флуд, но спрошу здесь пока помню.
Как видно на картинке, я сгруппировал функции (как аналогичные папки в Спотифае).

Чтоб быть уверенным что из "папки" у меня запустилась функция, которая мне нужна, есть более простой способ, нежели как вот так вот я придумал?...

```js
            function call(name, hasArgs) {
                let values = collectArgs(name, hasArgs);

                let funcName = '';
                if (values.hasOwnProperty('select')) {
                    funcName = values.select.name;
                } else {funcName = name;}
                
                google.script.run.withSuccessHandler(onSuccess).withFailureHandler(onError).withUserObject(funcName)[name](values)
                console.log(getTime(), 'Запущено', funcName);
 
            }
```

**Reply by Chimildic** [reply]
```js
console.log(getTime(), 'Запущено', values.select ? values.select.name : name);
```

**Reply by efremych** [reply]
> ```js
> console.log(getTime(), 'Запущено', values.select ? values.select.name : name);
> ```

Про эту конструкцию вспоминал, но недовспомнил. 
А "Завершено"?
Я с "Завершено" больше бодался. Разные имена фунций в "Запущено" и "Завершено" - не красиво оставлять)

А, всё, да, понял, эту конструкцию и вместо name вставить можно
```js
                google.script.run.withSuccessHandler(onSuccess).withFailureHandler(onError).withUserObject(values.select ? values.select.name : name)[name](values)
                console.log(getTime(), 'Запущено', values.select ? values.select.name : name);
```

Спасибо большое

**Reply by Chimildic** [reply]
Завершающая функция узнает имя отсюда же. Надо в `withUserObject` передать его. 
Наверное, будет некорректно если `select` не подразумевает функцию. Проверим и `name`. Остается его не использовать там, где нет вызова функции. Чуть изменил:
```js
function call(name, hasArgs) {
    let values = collectArgs(name, hasArgs);
    let customName = values.select && values.select.name ? values.select.name : name;
    google.script.run.withSuccessHandler(onSuccess).withFailureHandler(onError).withUserObject(customName)[name](values);
    console.log(getTime(), 'Запущено', customName);
}
```

**Reply by efremych** [reply]
Да, спасибо.

Не факт, конечно, что структура "папок" мне не надоест. Но при текущем раскладе для каждой "папки" надо создавать свою функцию или это тоже как-то можно сделать более красиво?

```js
function runFunction(values){
    this[values.select.name]();
}
```

Подозреваю, что дерево `const func `станет больше и сложнее и проще копипастить `runFunction`

**Reply by Chimildic** [reply]
Насколько вижу, все списки без передачи аргумента. Поэтому ничего создавать не надо. Просто во всех указывать `runFunction`. Или вопрос в другом?

**Reply by efremych** [reply]
Когда у меня были одинаковые названия функций, то у меня запускалась не та функция, которую я выбирал селектом. 
Поэтому я дал каждой уникальное имя и соответственно накопипастил одноименные «runFunction». 
Но теперь начал сомневаться :)

**Reply by Chimildic** [reply]
Возможно баг с генерацией id, поэтому не тот select собирается, поправимо

Скорее даже фича) Посмотрю как упростить

**Comment by Chimildic** [comment]
Поиск дал некоторую [подсказку](https://stackoverflow.com/a/45286695/5894542). Попробуем воспользоваться.

- Очистите кэш браузера за несколько дней или за все время.
- Обновите содержание [launch.html](https://github.com/Chimildic/goofy/blob/main/addons/Запуск%20с%20телефона/launch.html). Предварительно скопируйте куда-нибудь свой `const func`, чтобы вернуть после обновления.
- Обновите развертывание, как описано в [конце поста](https://github.com/Chimildic/goofy/discussions/9#discussioncomment-517316).

**Reply by efremych** [reply]
Результаты те же.
По ссылке что-то про iframe написано, наслышан, что оно довольно капризное и может по-разному себя вести вплоть до версии ОС...

**Reply by Chimildic** [reply]
Попробуйте удалить следующую строку вначале `launch.html` и повторить те же шаги. Это сломает стиль страницы. Но для теста надо проверить до конца. 
```html
<link href="https://bootswatch.com/4/darkly/bootstrap.min.css" rel="stylesheet" />
```

**Reply by efremych** [reply]
Стиль изменился, да.
Но на основном телефоне так и не заработало.
Что интересно, достал из чулана древний айфон, зарядил - так на нем всё прогрузилось (и до и после удаления строчки). Может правда, версия аппарата/ios...

**Reply by Chimildic** [reply]
Напишите какая версия системы на этих телефонах

**Reply by efremych** [reply]
Не заработало на iPhone SE (2020), ios 14.6. 
Заработало на iPhone 5S, ios 12.5.1

**Reply by Chimildic** [reply]
Попробуем исключить звено Apps Script Сходите по адресу [chimildic.github.io/test-wiki](https://chimildic.github.io/test-wiki/). Должны увидеть описание Spotify на википедии.

**Reply by efremych** [reply]
Страница wiki открылась. 

А страница с goofy грузится до половины и сбрасывается загрузка. 
![325305F4-F4EA-43DB-8E04-2FCA71F1B447](https://user-images.githubusercontent.com/76212904/124545787-4dcddc00-de32-11eb-9bd2-e5b82a622153.png)

**Reply by Chimildic** [reply]
Логи что нибудь показывают в это время? `chrome://inspect`

**Reply by efremych** [reply]
Нет, страница и раньше так грузилась. Решил упомянуть это.

**Reply by Chimildic** [reply]
На последнем скриншоте скроллбар видно. Сверху страницы нет описаний ошибок?

**Reply by efremych** [reply]
Нет, пустая страница. На одну строку «мотается» страница, да

**Reply by Chimildic** [reply]
Зайдите в режиме инкогнито. Он хоть ругнется, что доступа нет?

Еще вариант, в [своем репозитории](https://github.com/efremych/spotify-goofy/blob/main/index.html) отредактировать (иконка карандаша)`index.html`. Вставить до `iframe` следующие строчки. Отобразит на странице логи, если они есть вообще.
```js
<div id="logger"></div>
<script> !(function (o) { (console.old = console.log), (console.log = function () { var n, e, l = '<div class="p-2">'; for (e = 0; e < arguments.length; e++) (l += '<span class="log-' + typeof (n = arguments[e]) + '">'), 'object' == typeof n && 'object' == typeof JSON && 'function' == typeof JSON.stringify ? (l += JSON.stringify(n)) : (l += n), (l += '</span> '); (o.innerHTML += l + '</div>'), console.old.apply(void 0, arguments); }); })(document.getElementById('logger')); </script>
```

**Reply by efremych** [reply]
Чуть сочнее разве что страничка стала или кажется
![70E8DFAA-049C-473B-9F43-A8398584CDC2](https://user-images.githubusercontent.com/76212904/124644017-72f63500-de9a-11eb-94f4-6815c2c6507d.png)


Режим инкогнито никак себя не проявляет.

**Reply by Chimildic** [reply]
Не в том месте лог поставили. В `body` надо.
```html
<body>
    <-- сюда -->
    <iframe src="https://script.google.com/macros/s/AKfycbzmAfVL_ozEP69vpYvMo3t1Qlc4orPfk7eV5rWT/exec"></iframe>
</body>
```

**Reply by efremych** [reply]
есть такое, первое слово что увидел.
Но, нет, теперь просто без изменений.

**Reply by Chimildic** [reply]
Нет блокировщиков рекламы? С мобильной сети, не wi-fi пробовали?

**Reply by efremych** [reply]
И мобильная сеть и впн включал, и телефон перезагрузил, сейчас еще по настройкам браузера полазил (нашел только блокировать/не блокировать всплывающие окна - не повлияло). Особо-то много настроек на айфоне не будет.
В принципе готов остановиться на том, что проблема где-то у меня - может один случай на сотню, иначе, почему на древнем айфоне работает

**Comment by Chimildic** [comment, marked as answer]
Постараюсь объяснить. Цель инструкции в том, чтобы получить читаемый адрес `*.github.io/*` и скрыть баннер `Это приложение создано другим пользователем, а не компанией Google`. Единственный рабочий вариант связан с элементом `iframe`. Он позволяет встроить на страницу содержание другого сайта. В лучшем случае его не следует использовать, но у нас нет другого выбора, чтобы выполнить цель.

Содержание `script.google.com` не может отобразиться на нашем `*.github.io/*` просто так. В целях безопасности оно блокируется. С нашей стороны настройками приватности, со стороны Google заголовками в запросах. Чтобы не было возможности манипулировать встроенной страницей у автора внешнего сервиса. Авторизация на внешних ресурсах как правило проходит через открытие новой вкладки или нового окна (не встроенных `iframe`). 

Когда проходим авторизацию в любом сервисе Google, в браузере сохраняются куки, по которым нас можно идентифицировать и не заставлять каждый раз логиниться. Куки может смотреть как создатель (на своих адресах `*.google.*`) так и сторонние сайты на других адресах. Другой типичный пример это отслеживание для показа релевантной рекламы.

С марта 2020 года, [iOS 13+ по умолчанию блокируют перекрестное отслеживание](https://webkit.org/blog/10218/full-third-party-cookie-blocking-and-more/) (сторонние куки или 3rd party cookies). Другие платформы пока этого не делают. Поэтому iOS 12, Windows и Android показывали страницу сразу, а iOS 13+ белый экран или сообщение о блокировке. То есть куки и страница входа заблокированы. Первое из-за настройки браузера, второе из-за безопасности гугла и прочих ресурсов (заголовки в запросах).

Чтобы разрешить сторонние куки, в настройках телефона найти Safari или другие браузеры и отключить "мешать перекрестному отслеживанию" (еще может называться "сторонними файлами cookie").

**Reply by efremych** [reply]
![129BF64E-88C2-49B8-9C0F-E0F732524D79](https://user-images.githubusercontent.com/76212904/125496691-adf28a88-a751-4ac3-9a13-4ad66e220a35.png)
![990BD61E-8964-481E-A380-3DCE95110512](https://user-images.githubusercontent.com/76212904/125496698-39788947-c112-4d2f-94b6-57a592c7ebe2.png)
О, круто, сработало. И у Хрома оказывается тоже такая настройка есть - и в нем ссылка открылась. Спасибо!
